### PR TITLE
CXXODM-47 Mirror collection semantics in model where appropriate

### PR DESCRIPTION
**Pull request by adamchel**

### CODE REVIEW

File: src/mongo_odm/model.hpp [hunk 1]
```diff
@@ -35,6 +35,25 @@ struct FirstTypeIsTheSame : public std::false_type {};
...
 struct FirstTypeIsTheSame<T, T2, Ts...> : public std::is_same<T, std::decay_t<T2>> {};
 
+/**
+ * Type trait that checks whether or not a type is a container that contains a particular type.
+ *
+ * @tparam container_type The container being checked for containing a particular type.
+ * @tparam T The type that the container must hold for ContainerOfT::value to be true.
+ */
+template <typename container_type, typename T>
+struct ContainerOfT : public std::is_same<typename container_type::value_type, T> {};
```
Comment (Machyne) [inline src/mongo_odm/model.hpp]: can all of these type traits be moved into `utils.hpp` with the other type traits?

Also, for consistency, rename to snake case (and maybe `is_container_of`), and then offer `is_container_of_v`

File: src/mongo_odm/model.hpp [hunk 2]
```diff
@@ -35,6 +35,25 @@ struct FirstTypeIsTheSame : public std::false_type {};
...
+
+/**
+ * Type trait that checks whether or not an iterator iterates a particular type.
+ *
+ * @tparam iterator_type The iterator being checked for iterating a particular type.
+ * @tparam T The type that the iterator must iterate for IteratorOfT::value to be true.
+ */
+template <typename iterator_type, typename T>
+struct IteratorOfT
```
Comment (Machyne) [inline src/mongo_odm/model.hpp]: same

File: src/mongo_odm/model.hpp [hunk 3]
```diff
@@ -84,6 +160,133 @@ class model {
...
+     *
+     *  @tparam containter_type
+     *    The container type. Must contain an iterator that yields objects of this model.
+     *
+     *  @param container
+     *    Container of model objects to insert.
+     *  @param options
+     *    Optional arguments, see mongocxx::options::insert.
+     *
+     *  @return The result of attempting to performing the insert.
```
Comment (Machyne) [inline src/mongo_odm/model.hpp]: add `@see` links to [the insert docs](https://docs.mongodb.com/manual/tutorial/insert-documents/)

File: src/mongo_odm/query_builder.hpp [hunk 4]
```diff
@@ -292,7 +292,7 @@ class mod_expr {
      * as a key-value pair of the form " key: { $mod: [ divisor, remainder ] } "
 .
      * @param builder a BSON core builder
-     * @param wrap	    Whether to wrap the BSON inside a document.
+     * @param wrap      Whether to wrap the BSON inside a document.
```
Comment (Machyne) [inline src/mongo_odm/query_builder.hpp]: along as you're doing away with the tabs (here and above), can you fix the alignment?

File: src/mongo_odm/model.hpp (hunk 1 above)
Comment (adamchel) [inline src/mongo_odm/model.hpp]: Done.

File: src/mongo_odm/model.hpp (hunk 2 above)
Comment (adamchel) [inline src/mongo_odm/model.hpp]: Done.

File: src/mongo_odm/model.hpp (hunk 3 above)
Comment (adamchel) [inline src/mongo_odm/model.hpp]: Done.

File: src/mongo_odm/query_builder.hpp (hunk 4 above)
Comment (adamchel) [inline src/mongo_odm/query_builder.hpp]: Done.

File: src/mongo_odm/model.hpp [hunk 5]
```diff
@@ -56,13 +46,33 @@ class model {
...
+     *  @param options
+     *    Optional arguments, see mongocxx::options::count.
+     *
+     *  @return The count of the documents that matched the filter.
+     *  @throws mongocxx::exception::query if the count operation fails.
+     *
+     *  @see https://docs.mongodb.com/manual/reference/command/count/
+     */
+    static std::int64_t count(
+        bsoncxx::document::view_or_value filter,
```
Comment (Machyne) [inline src/mongo_odm/model.hpp]: I think `count` is a safe and useful place to default this option, but the C++11 driver doesn't do that. Thoughts? I would be happy to make this:
`bsoncxx::document::view_or_value filter = bsoncxx::document::view_or_value{}`

Comment (adamchel) [inline src/mongo_odm/model.hpp]: Done.

**Comment by Machyne** [comment]
LGTM

**Pull request merged**